### PR TITLE
Add Microsoft.AspNetCore.OpenApi package

### LINKED ISSUE #40676: Introduce Microsoft.AspNetCore.OpenApi package
Consider introducing a new package *Microsoft.AspNetCore.OpenApi* that provides integration features for ASP.NET Core HTTP APIs and OpenAPI. This package takes a dependency on [*Microsoft.OpenApi* ](https://github.com/microsoft/OpenAPI.NET) and contains the logic that maps primitive ASP.NET Core endpoint details to representations in endpoint metadata using the object model provided by *Microsoft.OpenApi*, along with projecting these details into the legacy `ApiExplorer` API from MVC. Furthermore, it provides helper methods for easily adding user-specified OpenAPI relevant details directly to endpoint metadata.

This would essentially be a strategy shift so that instead of continuing to add more metadata types to the framework to allow more details about endpoint APIs to be declared and published in MVC's `ApiExplorer` (e.g. #40084), we embrace the existing OpenAPI object model in *Microsoft.OpenApi* and allow all operation details to be specified in an endpoint's metadata and projected into `ApiExplorer` so that existing libraries like *Swashbuckle* continue to work.

The package would do the following:
- Depend on *Microsoft.OpenApi*
- Contain the logic that maps the `MethodInfo` for a route handler delegate in the endpoint metadata, to the requisite `Microsoft.OpenApi.Models.OpenApiOperation`
- Provide extension methods to easily allow mutating a route handler delegate's OpenApi metadata, e.g.
    ``` c#
    app.MapGet("/todos", async (TodoDb db) => await db.Todos.ToListAsync())
        .WithOpenApiDetails(operation =>
        {
            operation.OperationId = "GetAllTodos";
            operation.Description = "This API returns all the Todos.";
        });
    ```
- Preserves the `Microsoft.OpenApi.Models.OpenApiOperation` in the endpoint metadata
  - This requires a new extensibility point be added such that the package can mutate the metadata for endpoints built by the route endpoint builder
- Contain the `IApiDescriptionProvider` that projects a route handler's details from its OpenAPI endpoint metadata into `ApiExplorer`
- Be referenced in the Web API project template
- **Would not** reference *Swashbuckle.AspNetCore*, rather a project would still need to reference the SwaggerUI/OpenAPI library it wishes to use and call the requisite configuration methods for it to be configured

We need to consider of course how we do this without breaking (too much) stuff. In .NET 6, Minimal API endpoints are added to the `ApiExplorer` if MVC is configured, or the `AddEndpointsApiExplorer()` services extension method is called.

### PR DESCRIPTION
**Pull request by captainsafia**
Close https://github.com/dotnet/aspnetcore/issues/40676

- Adds new Microsoft.AspNetCore.OpenApi package to minimal templates
- Adds `ServiceProvider` to `EndpointBuilder` abstract class
- Adds `WithOpenApi()` extension method to `RouteHandlerBuilder`
- Adds `WithOpenApi(Func<OpenApiOperation, OpenApiOperation> modifyOperation)` extension method to `RouteHandlerBuilder`
- Adds stub `SchemaGenerator` to OpenApi package

### CODE REVIEW

File: src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        var serviceProvider = routeEndpointBuilder.ServiceProvider;
+
+        if (methodInfo == null || serviceProvider == null)
+        {
+            return null;
+        }
+
+        var hostEnvironment = serviceProvider.GetService<IHostEnvironment>();
+        var serviceProviderIsService = serviceProvider.GetService<IServiceProviderIsService>();
+        var generator = new OpenApiGenerator(hostEnvironment, serviceProviderIsService);
```
Comment (DamianEdwards) [inline src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs:82]: Now that `OpenApiGenerator` is no longer a singleton that's resolved from DI does it need to be refactored to make better use of its `ParameterBindingMethodCache` instance, etc.? e.g. `RequestDelegateFactory` creates a static instance of `ParameterBindingMethodCache`. That said, given this is in a package it might not be feasible to share the instance without making the type public anyway.

File: src/Http/OpenApi/src/OpenApiGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+                }
+                else
+                {
+                    requestBodyContent["application/json"] = new OpenApiMediaType
+                    {
+                        Schema = new OpenApiSchema
+                        {
+                            Type = SchemaGenerator.GetOpenApiSchemaType(requestBodyParameter.ParameterType)
+                        }
+                    }; ;
```
Comment (BrennanConroy) [inline src/Http/OpenApi/src/OpenApiGenerator.cs:314]:
```suggestion
                    };
```

File: src/Http/OpenApi/src/OpenApiGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+            {
+                Required = !isOptional,
+                Content = requestBodyContent
+            };
+        }
+
+        return null;
+    }
+
+    private IList<OpenApiTag> GetOperationTags(MethodInfo methodInfo, EndpointMetadataCollection metadata)
```
Comment (BrennanConroy) [inline src/Http/OpenApi/src/OpenApiGenerator.cs:335]: nit: Internal/private APIs should use `List` over `IList`

File: src/Http/OpenApi/src/OpenApiGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+                Required = !isOptional
+
+            };
+            openApiParameters.Add(openApiParameter);
+        }
+
+        return openApiParameters;
+    }
+
+    private static IDictionary<string, OpenApiMediaType> GetOpenApiParameterContent(EndpointMetadataCollection metadata)
```
Comment (BrennanConroy) [inline src/Http/OpenApi/src/OpenApiGenerator.cs:391]: Same here

File: src/Http/OpenApi/src/OpenApiGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+            return (true, null);
+        }
+        else if (parameter.CustomAttributes.Any(a => typeof(IFromServiceMetadata).IsAssignableFrom(a.AttributeType)) ||
+                parameter.ParameterType == typeof(HttpContext) ||
+                parameter.ParameterType == typeof(HttpRequest) ||
+                parameter.ParameterType == typeof(HttpResponse) ||
+                parameter.ParameterType == typeof(ClaimsPrincipal) ||
+                parameter.ParameterType == typeof(CancellationToken) ||
+                ParameterBindingMethodCache.HasBindAsyncMethod(parameter) ||
+                _serviceProviderIsService?.IsService(parameter.ParameterType) == true)
```
Comment (dougbu) [inline src/Http/OpenApi/src/OpenApiGenerator.cs:437]: nit
```suggestion
                _serviceProviderIsService?.IsService(parameter.ParameterType))
```

File: src/Http/OpenApi/src/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -0,0 +1,6 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.AspNetCore.OpenApi.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (dougbu) [inline src/Http/OpenApi/src/Properties/AssemblyInfo.cs:6]: Suggest getting rid of this file and adding the following to the project file:
``` msbuild
  <ItemGroup>
    <InternalsVisibleTo Include="Microsoft.AspNetCore.OpenApi.Tests" />
  </ItemGroup>
```

File: src/Http/OpenApi/src/SchemaGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.OpenApi;
+
+internal static class SchemaGenerator
```
Comment (dougbu) [inline src/Http/OpenApi/src/SchemaGenerator.cs:6]: Curious: in what sense is this a "schema generator"❔ It seems limited to returning a `string` categorizing a `Type`

File: src/Http/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests.csproj [hunk 8]
```diff
@@ -0,0 +1,25 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (dougbu) [inline src/Http/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests.csproj:5]: Why is this necessary❔

File: src/Http/OpenApi/test/OpenApiGeneratorTests.cs [hunk 9]
```diff
@@ -0,0 +1,814 @@
...
+        Assert.Null(operation);
+    }
+
+    [Fact]
+    public void UsesDeclaringTypeAsOperationTags()
+    {
+        var operation = GetOpenApiOperation(TestAction);
+
+        var declaringTypeName = typeof(OpenApiOperationGeneratorTests).Name;
+        var tag = Assert.Single(operation.Tags);
```
Comment (dougbu) [inline src/Http/OpenApi/test/OpenApiGeneratorTests.cs:35]: We usually separate Setup, Act and Assert rather than mix the first to in the middle block e.g.
```suggestion
        var declaringTypeName = typeof(OpenApiOperationGeneratorTests).Name;

        var tag = Assert.Single(operation.Tags);
```

File: src/Http/OpenApi/test/OpenApiGeneratorTests.cs [hunk 10]
```diff
@@ -0,0 +1,814 @@
...
+        {
+            var request = Assert.Single(operation.Parameters);
+            var content = Assert.Single(request.Content);
+            Assert.Equal("application/custom", content.Key);
+        }
+
+        AssertCustomRequestFormat(GetOpenApiOperation(
+            [Consumes("application/custom")]
+        (InferredJsonClass fromBody) =>
+            { }));
```
Comment (dougbu) [inline src/Http/OpenApi/test/OpenApiGeneratorTests.cs:65]: nit: This seems hard to read with the extra newlines. In this case, there's also an extra closing parenthesis.
```suggestion
        AssertCustomRequestFormat(GetOpenApiOperation(
            [Consumes("application/custom")] (InferredJsonClass fromBody) => { });
```

File: src/Http/HttpAbstractions.slnf [hunk 11]
```diff
@@ -40,18 +42,24 @@
       "src\\Http\\WebUtilities\\perf\\Microbenchmarks\\Microsoft.AspNetCore.WebUtilities.Microbenchmarks.csproj",
       "src\\Http\\WebUtilities\\src\\Microsoft.AspNetCore.WebUtilities.csproj",
       "src\\Http\\WebUtilities\\test\\Microsoft.AspNetCore.WebUtilities.Tests.csproj",
-      "src\\Http\\samples\\MinimalSample\\MinimalSample.csproj",
```
Comment (halter73) [inline src/Http/HttpAbstractions.slnf:43]: Can we leave this in the slnf?

File: src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /// <summary>
+    /// Adds an OpenAPI annotation to <see cref="Endpoint.Metadata" /> associated
+    /// with the current endpoint and modifies it with the given <paramref name="configureOperation"/>.
+    /// </summary>
+    /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
+    /// <param name="configureOperation">An <see cref="Action"/> that mutates an OpenAPI annotation.</param>
+    /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the endpoint.</returns>
+    public static RouteHandlerBuilder WithOpenApi(this RouteHandlerBuilder builder, Func<OpenApiOperation, OpenApiOperation> configureOperation)
```
Comment (halter73) [inline src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs:51]: Do we have any tests that call the WithOpenApi overloads? The ones I see all call into OpenApiGenerator directly.

File: src/Http/OpenApi/src/OpenApiGenerator.cs [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+        }
+
+        if (eligibileAnnotations.Count == 0)
+        {
+            GenerateDefaultResponses(eligibileAnnotations, responseType);
+        }
+
+        foreach (var annotation in eligibileAnnotations)
+        {
+            var statusCode = $"{annotation.Key}";
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/OpenApi/src/OpenApiGenerator.cs:193]: No interpolation needed. The status code needs to be stringified for use in the `OpenApiResponses` dictionary so I've updated this to a `ToString()`

File: src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        return builder;
+
+    }
+
+    /// <summary>
+    /// Adds an OpenAPI annotation to <see cref="Endpoint.Metadata" /> associated
+    /// with the current endpoint and modifies it with the given <paramref name="configureOperation"/>.
+    /// </summary>
+    /// <param name="builder">The <see cref="RouteHandlerBuilder"/>.</param>
+    /// <param name="configureOperation">An <see cref="Action"/> that mutates an OpenAPI annotation.</param>
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs:49]: Good catch! Fixed this to reflect the new API.

File: src/Http/OpenApi/src/OpenApiGenerator.cs [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+    private static OpenApiResponses GetOpenApiResponses(MethodInfo method, EndpointMetadataCollection metadata)
+    {
+        var responses = new OpenApiResponses();
+        var responseType = method.ReturnType;
+        if (AwaitableInfo.IsTypeAwaitable(responseType, out var awaitableInfo))
+        {
+            responseType = awaitableInfo.ResultType;
+        }
+
+        if (typeof(IResult).IsAssignableFrom(responseType))
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/OpenApi/src/OpenApiGenerator.cs:103]: In the long-term, I'd like the simplify the logic for producing responses in general. At the moment, this is carrying over logic from the `EndpointMetadataApiDescriptionProvider` which in turn copies some patterns to MVC. I think there's a simpler collection of rules that we can create here but we'll want to approach that later.

Filed https://github.com/dotnet/aspnetcore/issues/41243 to track this.

File: src/Http/OpenApi/src/OpenApiGenerator.cs (hunk 2 above)
Comment (captainsafia) [reply src/Http/OpenApi/src/OpenApiGenerator.cs:314]: 🙊

File: src/Http/OpenApi/src/OpenApiGenerator.cs (hunk 3 above)
Comment (captainsafia) [reply src/Http/OpenApi/src/OpenApiGenerator.cs:335]: TIL. Fixed.

File: src/Http/OpenApi/src/Microsoft.AspNetCore.OpenApi.csproj [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    <PackageTags>aspnetcore;openapi</PackageTags>
+    <Description>Provides APIs for annotating route handler endpoints in ASP.NET Core with OpenAPI annotations.</Description>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Reference Include="Microsoft.OpenApi" />
+    <Reference Include="Microsoft.AspNetCore.Http.Abstractions" />
+    <Reference Include="Microsoft.AspNetCore.Routing" />
+    <Reference Include="Microsoft.AspNetCore.Mvc.Core" />
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/OpenApi/src/Microsoft.AspNetCore.OpenApi.csproj:15]: I posted a response on the other PR that I closed.

The requirement for a dependency on MVC largely stems from some carry-over in the way we generated Response annotations in this library. Moving forward, we want to simplify this logic (see https://github.com/dotnet/aspnetcore/issues/41243) and remove the dependency on the types defined in MVC.

As I typed this out, I realized that we also have a long-term plan to include methods for converting `OpenApiOperation` back to MVC types for backwards compat, although I'm not enthused by those plans anymore.

File: src/Http/OpenApi/src/OpenApiGenerator.cs [hunk 17]
```diff
@@ -0,0 +1,472 @@
...
+namespace Microsoft.AspNetCore.OpenApi;
+
+/// <summary>
+/// Defines a set of methods for generating OpenAPI definitions for endpoints.
+/// </summary>
+internal class OpenApiGenerator
+{
+    private readonly IHostEnvironment? _environment;
+    private readonly IServiceProviderIsService? _serviceProviderIsService;
+    private readonly ParameterBindingMethodCache ParameterBindingMethodCache = new();
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/OpenApi/src/OpenApiGenerator.cs:30]: This should actually be a `private static readonly` per the convention [here](https://github.com/dotnet/aspnetcore/blob/7a1a1ebf86fb24d84cefd45e05534c79708c7947/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L34).

File: src/Http/OpenApi/src/OpenApiGenerator.cs (hunk 5 above)
Comment (captainsafia) [reply src/Http/OpenApi/src/OpenApiGenerator.cs:437]: `?.IsService` will return a `bool?` so we need to do an explicit check to get a `bool` here.

File: src/Http/OpenApi/src/OpenApiGenerator.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/Http/OpenApi/src/OpenApiGenerator.cs:103]: I think with Damians Result.Typed change, the appropriate metadata will be added for IResults (assuming you're using the new APIs of course). So this scenario will be mostly dealt with.

File: src/Http/OpenApi/src/Properties/AssemblyInfo.cs (hunk 6 above)
Comment (captainsafia) [reply src/Http/OpenApi/src/Properties/AssemblyInfo.cs:6]: Good point!

This reminded me of an observation that I had made earlier about inconsistencies between IVT being specified via `AssemblyInfo` and IVT being configured via build props.

I opened https://github.com/dotnet/aspnetcore/issues/41244 to track us being consistent through the repo here in the event we want to take a stand a particular approach.

File: src/Http/OpenApi/src/SchemaGenerator.cs (hunk 7 above)
Comment (captainsafia) [reply src/Http/OpenApi/src/SchemaGenerator.cs:6]: It's a reductive implementation for something we plan on improving in an upcoming preview (see https://github.com/dotnet/aspnetcore/issues/41246).

A lot of the work for mapping .NET types over to JSON validation schemas is done in Swashbuckle's Swagger generator at the moment. Since we're adopting first-class support for OpenApi in this package, we want to roll out our own schema generator built-in.

File: src/Http/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests.csproj (hunk 8 above)
Comment (captainsafia) [reply src/Http/OpenApi/test/Microsoft.AspNetCore.OpenApi.Tests.csproj:5]: Copy-pasta. 🙊

File: src/Http/OpenApi/test/OpenApiGeneratorTests.cs (hunk 9 above)
Comment (captainsafia) [reply src/Http/OpenApi/test/OpenApiGeneratorTests.cs:35]: Good catch!

File: src/Http/OpenApi/test/OpenApiGeneratorTests.cs (hunk 10 above)
Comment (captainsafia) [reply src/Http/OpenApi/test/OpenApiGeneratorTests.cs:65]: Fixed this up and some other places where inline lambdas with attributes were a bit askew.

File: src/Http/HttpAbstractions.slnf (hunk 11 above)
Comment (captainsafia) [reply src/Http/HttpAbstractions.slnf:43]: Brought this back. Surprisingly, the build in VS works now so I assume it was including the other properties that was required.

File: src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs (hunk 12 above)
Comment (captainsafia) [reply src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs:50]: Added some test coverage here.

File: src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs (hunk 1 above)
Comment (captainsafia) [reply src/Http/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs:81]: Yeah, the addition of this package certainly ups the case for https://github.com/dotnet/aspnetcore/issues/33644 and introduces the additional challenge of this particular codepath needing to return `ParameterLocation` instead of `BindingSource`.

At least for all the complex types, the new `IProvidesMetadata` stuff might help a bit but not sure what to do about primitives.

File: src/Http/OpenApi/src/Microsoft.AspNetCore.OpenApi.csproj (hunk 16 above)
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/OpenApi/src/Microsoft.AspNetCore.OpenApi.csproj:15]: @davidfowl mentioned that it might considering moving this to its own top-level directory under the `src` repo. That feels a little too heavy duty for me but might make the most sense considering this is not part of the framework anyways (e.g. avoid assuming it is a primitive like Http is).

[the remark this answers is not in the document]
Comment (brunolins16) [reply src/Http/OpenApi/src/Microsoft.AspNetCore.OpenApi.csproj:15]: Moving to a top-level directory might also help to remove the MVC projects from the `HttpAbstractions.slnf`, right?

[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Http/OpenApi/src/Microsoft.AspNetCore.OpenApi.csproj]: OK. I've moved it to a top-level directory.

File: src/OpenApi/src/Microsoft.AspNetCore.OpenApi.csproj [hunk 18]
```diff
@@ -0,0 +1,29 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>aspnetcore;openapi</PackageTags>
+    <Description>Provides APIs for annotating route handler endpoints in ASP.NET Core with OpenAPI annotations.</Description>
+    <Nullable>enable</Nullable>
```
Comment (BrennanConroy) [inline src/OpenApi/src/Microsoft.AspNetCore.OpenApi.csproj:8]: nit: not needed as it's the default now

File: src/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+                var openApiOperation = GetOperationForEndpoint(routeEndpointBuilder);
+                if (openApiOperation != null)
+                {
+                    routeEndpointBuilder.Metadata.Add(openApiOperation);
+                }
+            };
+        });
+        return builder;
+
+    }
```
Comment (BrennanConroy) [inline src/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs:41]:
```suggestion

        return builder;
    }
```

File: src/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+        builder.Add(endpointBuilder =>
+        {
+            if (endpointBuilder is RouteEndpointBuilder routeEndpointBuilder)
+            {
+                var openApiOperation = GetOperationForEndpoint(routeEndpointBuilder);
+                if (openApiOperation != null)
+                {
+                    routeEndpointBuilder.Metadata.Add(configureOperation(openApiOperation));
+                }
+
```
Comment (BrennanConroy) [inline src/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs:61]:
```suggestion
```

File: src/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+            };
+        });
+        return builder;
```
Comment (BrennanConroy) [inline src/OpenApi/src/OpenApiRouteHandlerBuilderExtensions.cs:64]:
```suggestion

        return builder;
```

File: src/OpenApi/src/OpenApiGenerator.cs [hunk 22]
```diff
@@ -0,0 +1,474 @@
...
+internal class OpenApiGenerator
+{
+    private readonly IHostEnvironment? _environment;
+    private readonly IServiceProviderIsService? _serviceProviderIsService;
+
+    private static readonly ParameterBindingMethodCache ParameterBindingMethodCache = new();
+
+    /// <summary>
+    /// Creates an <see cref="OpenApiGenerator" /> instance given a <see cref="IHostEnvironment" />
+    /// and a <see cref="IServiceProviderIsService" /> instance.
```
Comment (BrennanConroy) [inline src/OpenApi/src/OpenApiGenerator.cs:36]:
```suggestion
    /// Creates an <see cref="OpenApiGenerator" /> instance given an <see cref="IHostEnvironment" />
    /// and an <see cref="IServiceProviderIsService" /> instance.
```

File: src/OpenApi/src/OpenApiGenerator.cs [hunk 23]
```diff
@@ -0,0 +1,474 @@
...
+    /// </summary>
+    /// <param name="environment">The host environment.</param>
+    /// <param name="serviceProviderIsService">The service to determine if the a type is available from the <see cref="IServiceProvider"/>.</param>
```
Comment (BrennanConroy) [inline src/OpenApi/src/OpenApiGenerator.cs:39]:
```suggestion
    /// <param name="serviceProviderIsService">The service to determine if the type is available from the <see cref="IServiceProvider"/>.</param>
```

File: src/OpenApi/src/OpenApiGenerator.cs [hunk 24]
```diff
@@ -0,0 +1,474 @@
...
+            OperationId = metadata.GetMetadata<IEndpointNameMetadata>()?.EndpointName,
+            Summary = metadata.GetMetadata<IEndpointSummaryMetadata>()?.Summary,
+            Description = metadata.GetMetadata<IEndpointDescriptionMetadata>()?.Description,
+            Tags = GetOperationTags(methodInfo, metadata),
+            Parameters = GetOpenApiParameters(methodInfo, metadata, pattern, disableInferredBody),
+            RequestBody = GetOpenApiRequestBody(methodInfo, metadata, pattern),
+            Responses = GetOpenApiResponses(methodInfo, metadata)
+        };
+
+        static bool ShouldDisableInferredBody(string method)
```
Comment (BrennanConroy) [inline src/OpenApi/src/OpenApiGenerator.cs:84]: super nit: should this be in a shared location?

File: src/OpenApi/src/OpenApiGenerator.cs [hunk 25]
```diff
@@ -0,0 +1,474 @@
...
+
+            var nullabilityContext = new NullabilityInfoContext();
+            var nullability = nullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState != NullabilityState.NotNull;
+            var openApiParameter = new OpenApiParameter()
+            {
+                Name = parameter.Name,
+                In = parameterLocation,
+                Content = GetOpenApiParameterContent(metadata),
+                Schema = new OpenApiSchema { Type = SchemaGenerator.GetOpenApiSchemaType(parameter.ParameterType) },
```
Comment (BrennanConroy) [inline src/OpenApi/src/OpenApiGenerator.cs:383]: This should take TryParse into account, right?
https://github.com/dotnet/aspnetcore/blob/64ccf823287a1de02385d29497127ad5197a389d/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L284-L286

File: src/OpenApi/src/OpenApiGenerator.cs [hunk 26]
```diff
@@ -0,0 +1,474 @@
...
+            foreach (var contentType in acceptsMetadata.ContentTypes)
+            {
+                openApiParameterContent.Add(contentType, new OpenApiMediaType());
+            }
+        }
+
+        return openApiParameterContent;
+    }
+
+    private (bool, ParameterLocation?) GetOpenApiParameterLocation(ParameterInfo parameter, RoutePattern pattern, bool disableInferredBody)
```
Comment (BrennanConroy) [inline src/OpenApi/src/OpenApiGenerator.cs:408]: Could you name the tuple items? Makes it easier to track what they are.

File: src/ProjectTemplates/Web.ProjectTemplates/content/WebApi-CSharp/Program.cs [hunk 27]
```diff
@@ -5,6 +5,9 @@
 #if (WindowsAuth)
 using Microsoft.AspNetCore.Authentication.Negotiate;
 #endif
+#if (EnableOpenAPI)
+using Microsoft.AspNetCore.OpenApi;
```
Comment (BrennanConroy) [inline src/ProjectTemplates/Web.ProjectTemplates/content/WebApi-CSharp/Program.cs:9]: Do we normally add `using`s without adding code that uses the `using`?

File: src/OpenApi/src/OpenApiGenerator.cs (hunk 25 above)
Comment (captainsafia) [reply src/OpenApi/src/OpenApiGenerator.cs:383]: The rule has to factor in the underlying type of the type with the `TryParse` so the check is a little more involved with that. Added a callout to it in https://github.com/dotnet/aspnetcore/issues/41246

File: src/ProjectTemplates/Web.ProjectTemplates/content/WebApi-CSharp/Program.cs (hunk 27 above)
Comment (captainsafia) [reply src/ProjectTemplates/Web.ProjectTemplates/content/WebApi-CSharp/Program.cs:9]: Good point! I added this as a nudge to the existence of the package and for compat with the other `Program.cs` but it's probably find to omit it for a first round.

EDIT: I also just realized this is the modern day controllers template, not the minimal template. Removed.

File: src/OpenApi/src/OpenApiGenerator.cs (hunk 26 above)
Comment (captainsafia) [reply src/OpenApi/src/OpenApiGenerator.cs]: Good call!

File: src/OpenApi/src/OpenApiGenerator.cs (hunk 24 above)
Comment (captainsafia) [reply src/OpenApi/src/OpenApiGenerator.cs:84]: Meh, considering how small it is I think it's almost not worth the effort of putting it in a shared location/adding includes/etc.

File: src/OpenApi/src/OpenApiGenerator.cs [hunk 28]
```diff
@@ -0,0 +1,474 @@
...
+        var eligibileAnnotations = new Dictionary<int, (Type?, MediaTypeCollection)>();
+
+        foreach (var responseMetadata in producesResponseMetadata)
+        {
+            var statusCode = responseMetadata.StatusCode;
+
+            var discoveredTypeAnnotation = responseMetadata.Type;
+            var discoveredContentTypeAnnotation = new MediaTypeCollection();
+
+            if (discoveredTypeAnnotation == typeof(void))
```
Comment (BrennanConroy) [inline src/OpenApi/src/OpenApiGenerator.cs:125]: Is this needed? Looks like we check null or void shortly below and set to responseType then.

**Review by BrennanConroy — APPROVED** [review]
Would like someone else to look at the PR too since I'm only passingly familiar with the area. But looks good overall 😃

**Review by halter73 — APPROVED** [review]

File: src/OpenApi/src/OpenApiGenerator.cs (hunk 28 above)
Comment (captainsafia) [reply src/OpenApi/src/OpenApiGenerator.cs:125]: Yeah, this isn't needed. It exists as a legacy from the ApiExplorer days where `== typeof(void)` was used to handle scenarios where the return type was not provided by the user via an attribute, in which case, we would only fall back to the defined responseType on a 2xx status code.

However, that was coupled with some weirdness in the way that content types were handled in 

Anyhoo, I was going to clean this up in a follow-up commit had I gotten any feedback in addition to this on the PR but I'll do this as part of https://github.com/dotnet/aspnetcore/issues/41243 instead.
